### ISSUE TITLE
saving my continuous progress so i won't have to start over.

### ISSUE THREAD

**Issue by DR2495** [SkalskiP/make-sense#169]
saving my continuous progress so I won't have to start over.

**Comment by iamfaith** [comment]
I suggest make-sense should add the concept of projects. We can click ```new project``` to start our work and save for next time.

**Comment by rasyidf** [comment]
Hi @SkalskiP , I found something interesting here :
[Paint.js](https://github.com/christianliebel/paint)

**Comment by SkalskiP** [comment]
Hello, once again @rasyidf  👋. That is a true, very interesting project. But... what is the relation of paint.js with that issue? 🧐

**Comment by rasyidf** [comment]
I guess I missed the details before,

If we want to fully utilize the PWA capability to just install the webapp locally, then integrate file handler. 

[Web.dev file handling](https://web.dev/file-handling/)
 
we,  as a user can

- start project
- do stuff
- save as myproject.mks (for example)
- close the app

on our desktop,

- click myproject.mks.
- resume works.

**Comment by SkalskiP** [comment]
@rasyidf I was thinking about it. We could do something similar. But to make it work similarly - load a whole project from a single file, we would need that file to be quite big. We would need to use some kind of zip structure and hide all images inside along with labels and metadata required.

**Comment by rasyidf** [comment]
I see, so basically you want the project behave like open Document format. read here => [Wikipedia Open Doc.](https://en.m.wikipedia.org/wiki/OpenDocument)  

make a zipped folder, with Metadata and every assets.

**Comment by SkalskiP** [comment]
Yes but instead of using XML we would use JSON to store labels and metadata. We are already saving zip in TypeScript so I do not think it would be an issue. One thing that is worrying is the size of the saved file. It can be easily a few hundred MB. So for some users with slower machines it can be an issue :/

**Comment by rasyidf** [comment]
If the make-sense is installed offline. how about making it folder based project with something like .sense/ subfolder with it's workspaced settings.  it's opinionated but can be easily predicted by program

something like this : project.json
```json
{
    "name": "pet-behaviour",
    "version": "2.3.0",
    "description": "Pet Behaviour Recognition",
    "keywords": ["veterinary", "cats", "pets", "dogs", "YOLOv5"],
    "author": "John Doe",
    "engine": {
      "model": "YOLOv5",
      "weight" : "yolov5.weights"
    },
    "paths": {
      "images": [
        "images/*"
      ],
      "annotations": [
        "labels/*"
      ],
      "masks": [
        "masks/*"
      ],
      "code": [
        "src/*"
      ]
    }
  }
```

then, we can make the folder structured as described, then load it and the images will be lazy loaded, after every setup finalized.

**Comment by SkalskiP** [comment]
@rasyidf that's an interesting suggestion, but I really wouldn't like to have a separate solution for people using make-sense online and offline. We have around 6000 people visiting https://www.makesense.ai/ every week. And I think that we should build new features mostly for those people.

**Comment by AxelAli** [comment]
I believe we could use either indexedDB (similar to MySQL) or  https://developer.mozilla.org/em/docs/Web/API/Window/localStorage.
We could store the pending annotations by item name/hash into indexedDB and that would persist by the user side without the need for any storage on our servers (this being particularly important in our client-side system)

**Comment by SkalskiP** [comment]
Hello, @AxelAli 👋  ! Many thanks for your interest in our project and for your ideas. However, both of them have some drawbacks that will probably stop us from using them. 
- `DB` - To minimize the cost MakeSense is a purely frontend-based solution. Consequently, the proposed solution must work without a backend. The DB would have to be hosted somewhere which would result in a cost to the project.
- `localStorage` - It will only allow us to store labels information. We would like to store the entire project, including photos.

**Comment by AxelAli** [comment]
Hi @SkalskiP , IndexedDB lives in the user's browser, no installs required.
Here you can see an example [https://mdn.github.io/learning-area/javascript/apis/client-side-storage/indexeddb/notes/](demo) (Try adding a note, changing the URL to Google or something, and open the same link in another tab), notes will persist.

**Comment by SkalskiP** [comment]
Hi @AxelAli! I have to admit that these are very interesting ideas. 💪  I'll have to look into that. I'm curious to see how efficient this solution is. We have to remember that projects have very often 500 or 1000 images. But thank you very much for this idea. I am also sorry that I reacted so negatively. I was simply not aware of the existence of this feature. 🙏

Let me do the research.
